iotileemulate/iotile/emulate/virtual/emulated_device.py

```diff
@@ -46,31 +43,117 @@ def __init__(self, iotile_id, name):
 
         self._logger = logging.getLogger(__name__)
         self._rpc_queue = WorkQueueThread(self._background_dispatch_rpc)
+        self._pending_rpcs = {}
+        self._deadlock_check = threading.local()
+        self._setup_deadlock_check()
+
+    def _setup_deadlock_check(self):
+        """Initialize thread-local storage to detect potential deadlocks.
+
+        If a blocking RPC is called from the rpc thread itself, then it can
+        never complete because the thread that can execute it is blocked
+        waiting for it to finish.  This can lead to tough-to-debug situations
+        so we automatically detect when this happens and throw an exception
+        instead.
+
+        The check is based on creating a thread-local variable that is true
+        in the rpc dispatch thread and false in all other threads.  This
+        variable is checked before queuing an RPC.  If it is true, then
+        we would deadlock and an exception is thrown instead.
+
+        Since the rpc thread isn't running until start() is called, we
+        queue a deferred task to setup the thread local as the first
+        thing done by the background thread.
+        """
+
+        self._deadlock_check.__dict__.setdefault('is_rpc_thread', False)
+
+        def _init_tls():
+            self._deadlock_check.is_rpc_thread = True
+
+        self._rpc_queue.defer(_init_tls)
 
     def _background_dispatch_rpc(self, action):
         """Background work queue handler to dispatch RPCs."""
 
         address, rpc_id, arg_payload = action
 
-        try:
-            exc_status = None
-            resp = None
+        # FIXME: Check for a queued RPC and return busy
```

Comment:
same

Reply:
Converted to an issue since the feature is not trivial #619.